src/main/scala/mesosphere/marathon/state/MarathonStore.scala

```diff
@@ -1,139 +1,75 @@
-            }
-            None
-        }
-      case None =>
-        throw new StorageException(s"Failed to read $key, does not exist, should have been created automatically.")
+    store.load(prefix + key).map {
+      _.map{ entity =>
+        bytesRead.update(entity.bytes.length)
+        stateFromBytes(entity.bytes)
+      }
     }
   }
 
-  def modify(key: String)(f: (() => S) => S): Future[Option[S]] = {
-    val lock = locks.get(key)
-    lock.acquire()
-
-    val res: Future[Option[S]] = state.fetch(prefix + key) flatMap {
-      case Some(variable) =>
-        bytesRead.update(variable.value.length)
-        val deserialize = { () =>
-          try {
-            stateFromBytes(variable.value)
-          }
-          catch {
-            case e: InvalidProtocolBufferException =>
-              if (variable.value.nonEmpty) {
-                log.error(s"Failed to read $key, could not deserialize data (${variable.value.length} bytes).", e)
-              }
-              newState()
-          }
-        }
-        val newValue: S = f(deserialize)
-        state.store(variable.mutate(newValue.toProtoByteArray)) map {
-          case Some(newVar) =>
-            bytesWritten.update(newVar.value.size)
-            Some(stateFromBytes(newVar.value))
-          case None =>
-            throw new StorageException(s"Failed to store $key")
-        }
+  def modify(key: String)(f: (() => S) => S): Future[S] = withLock(key) {
+    val res = store.load(prefix + key).flatMap {
+      case Some(entity) =>
+        bytesRead.update(entity.bytes.length)
+        val updated = f(() => stateFromBytes(entity.bytes))
+        val updatedEntity = entity.mutate(updated.toProtoByteArray)
+        bytesWritten.update(updatedEntity.bytes.length)
+        store.save(updatedEntity)
       case None =>
-        throw new StorageException(s"Failed to read $key, does not exist, should have been created automatically.")
+        val created = f(() => newState()).toProtoByteArray
+        bytesWritten.update(created.length)
+        store.create(prefix + key, created)
     }
-
-    res onComplete { _ =>
-      lock.release()
-    }
-
-    res
+    res.map { entity => stateFromBytes(entity.bytes) }
   }
 
-  def expunge(key: String): Future[Boolean] = {
-    val lock = locks.get(key)
-    lock.acquire()
-
-    val res = state.fetch(prefix + key) flatMap {
-      case Some(variable) =>
-        bytesRead.update(Option(variable.value).map(_.length).getOrElse(0))
-        state.expunge(variable) map {
-          case Some(b) => b.booleanValue()
-          case None    => throw new StorageException(s"Failed to expunge $key")
-        }
-
-      case None => throw new StorageException(s"Failed to read $key")
-    }
-
-    res onComplete { _ =>
-      lock.release()
-    }
-
-    res
+  def expunge(key: String): Future[Boolean] = withLock(key) {
+    store.delete(prefix + key) map { _ => true }
```

Comment:
Sounds good. The only thing that is missing then is the ability to take a look at the old instance (since we retrieve it anyway). But since we do not need that...

Besides: `Success(None)` is nothing like `Some(null)` but may make a lot of sense. `Future[Option[...]]` is common and useful.